3rd & 4th chapters of Nest GS (#6)
raven-os/getting-started: nest/03.advanced_concepts.md

# 3 Advanced concepts
In this chapter, we will detail some advanced concepts related to `nest`, which will help understanding how `nest` works, but aren't strictly needed for a daily use of `nest`.

> On the other hand, readers interested in `finest` should read this chapter, because those concepts are important to understand how `finest` works.

Reviewer: "On the other hand" is not used properly. The first and second sentences don't form pros & cons.

## 3.1 Managing repositories
### 3.1.1 The official repositories
The official repositories are the ones maintained by the Raven-OS team.
For now, only one is available: `stable`, which is the one enabled by default when installing Nest.
It contains all the stable and safe-to-use packages for Nest, and is maintained by the Raven-OS team.

### 3.1.2 The unofficial repositories
There are also unofficial repositories for Nest.

> They are not being maintained by the Raven-OS team or someone trusted by the team.

Reviewer: ```suggestion
These repositories are not vetted by Raven-OS, so caution is advised.
```
This sentence implies a strong negative, when in reality such a repo could have a relation with someone from Raven-OS.

> You must decide whether to trust their maintainers and take full responsibility for any consequences of using any unofficial repository.

Reviewer: This may imply we take some kind of responsibility for the official repository. I don't like this sentence, it may even give us legal trouble. I'd remove it or change it to this:
```suggestion
It is up to you to decide whether to trust their maintainers.
```

### 3.1.3 Configuring repositories
The repositories configuration is in the `config.toml` file of Nest.

> It would look something like this:

Reviewer: ```suggestion
It should ressemble something like this:
```
(I don't understand the sentence's meaning with would)

```toml
# Stable repository
[repositories.stable]
mirrors = ["https://stable.raven-os.org"]

# Beta repository (uncomment to enable)
# [repositories.beta]
# mirrors = ["https://beta.raven-os.org"]

# Nightly repository (uncomment to enable)
# [repositories.nightly]
# mirrors = ["https://nightly.raven-os.org"]
```

> A mirror is basically a copy of all datas from the main server.

Reviewer: ```suggestion
A mirror is basically a copy of all the data on the main server.
```
"Data" is weird when it comes to singular / plural, it is already plural in a sense.

They exist for redundancy and speed.

> Mirrors are spread all over the globe in strategic locations.

Reviewer: Is this actually true? :P

Author: ```suggestion
Mirrors are *supposedly* spread all over the globe in strategic locations.
```

Reviewer: xD

If you live in Sweden and use a mirror from Sweden, the download speed should be faster than if you used a mirror from Australia for example.
The point of mirrors is to choose the nearest one from your location, so you'll download packages as fast as possible.

> It's advised to have several mirrors for each repository Nest uses, because a mirror can be down or broken for a small period of time, or just be down forever.

Reviewer: I don't have a better formulation in my mind right now, but this does sound clunky to me.

Author: ```suggestion
Because a mirror can be down or broken for a short period of time or just forever, it's advised to have more than one mirror for each repository used by Nest.
```

Reviewer: (edited your comment to fix typo)

If Nest can't reach a mirror, it will try with the next one in the list.

Only `stable` is enabled by default, so you'll have to uncomment the other repositories and their mirrors in order to use them.

You can also add an unofficial repository to this list, by following the same syntax shown above in the code excerpt.

## 3.2 Overview of the cache

> Nest is using a cache to store a copy of the list of all available packages for Nest on the OS.

Reviewer: ```suggestion
Nest uses a cache to store a copy of the list of all available packages for Nest on the OS.
```

Nest can only install packages present in your cache.

> That's why you need to **keep it up-to-date as often as possible**, otherwise Nest won't install new versions of packages that came out on Nest's server.

Reviewer: Shouldn't it rather be "make sure it is up to date when doing an operation" or something like that?
Keeping it up to date when doing no operation seems useless.

There are 3 different folders for the cache:
* `/var/nest/available` contains the list of all available packages for Nest.
The content of this folder gets updated everytime you update your local cache with Nest.
* `/var/nest/downloaded` has all the packages that Nest has downloaded, whether they are installed or not.
If you installed a package before, and then uninstalled it, Nest will keep its archive there, in case you change your mind and want to install that same version of the same package later.
This allows Nest to speed up the process of installation, as it doesn't have to download the same archive again.
* `/var/nest/installed` contains all the packages that are currently installed on the OS.

## 3.3 Version requirement
[//]: # (TODO: add link to the section 1.4)
As explained in [the section 1.4](), the packages have a specific syntax regarding their name.

> A thing that wasn't mentionned at the time was that you can have requirement(s) on the version of the package you want to install.

Reviewer: I don't like the current wording of the section at all, I'd prefer something like this:

---

> "Sometimes you don't need a specific version of a package, but only that it fits certain prerequisites.
> Version requirements solve this problem by allowing you to specify a minimum or maximum version.
> 
> Examples:[...]"

Followed by examples of course.

---
The problem with the current text is that it doesn't present a reason to use this and explains nothing (you have to understand from the examples). It may be a good idea and go even further by explicitly giving the version requirement syntax.

Author: I'm giving examples right after.It feels redundant to me to explicitly give or explain the version requirement syntax. Like should I explain all the operators you can use for the version requirement syntax?

Reviewer: It may sound stupid, but "<", "<=", "==", ">=", ">" are not necessarily the symbols someone may expect (Ex: I'm not sure if it's "=" or "=="). The order is also important, it might be a good idea to talk about semver.

# 3 Advanced concepts
In this chapter, we will detail some advanced concepts related to `nest`, which will help understanding how `nest` works, but aren't strictly needed for a daily use of `nest`.
Readers interested in `finest` should read this chapter, because those concepts are important to understand how `finest` works.

## 3.1 Managing repositories
### 3.1.1 The official repositories
The official repositories are the ones maintained by the Raven-OS team.
For now, only one is available: `stable`, which is the one enabled by default when installing Nest.
It contains all the stable and safe-to-use packages for Nest, and is maintained by the Raven-OS team.

### 3.1.2 The unofficial repositories
There are also unofficial repositories for Nest.
These repositories are not vetted by Raven-OS, so caution is advised.
It is up to you to decide whether to trust their maintainers.

### 3.1.3 Configuring repositories

> The repositories configuration is in the `config.toml` file of Nest.

Reviewer 2: ```suggestion
The repositories configuration is specified in the configuration file of Nest (`/etc/nest/config.toml`).
```

It should ressemble something like this:

```toml
# Stable repository
[repositories.stable]
mirrors = ["https://stable.raven-os.org"]

# Beta repository (uncomment to enable)
# [repositories.beta]
# mirrors = ["https://beta.raven-os.org"]

> # Nightly repository (uncomment to enable)

Reviewer 2: Nightly has been renamed to unstable.

# [repositories.nightly]
# mirrors = ["https://nightly.raven-os.org"]
```

A mirror is basically a copy of all data from the main server.
They exist for redundancy and speed.
Mirrors are spread all over the globe in strategic locations.
If you live in Sweden and use a mirror from Sweden, the download speed should be faster than if you used a mirror from Australia for example.
The point of mirrors is to choose the nearest one from your location, so you'll download packages as fast as possible.
Because a mirror can be down or broken for a short period of time or just forever, it's advised to have more than one mirror for each repository used by Nest.
If Nest can't reach a mirror, it will try with the next one in the list.

Only `stable` is enabled by default, so you'll have to uncomment the other repositories and their mirrors in order to use them.

You can also add an unofficial repository to this list, by following the same syntax shown above in the code excerpt.

## 3.2 Overview of the cache

> Nest uses a cache to store a copy of the list of all available packages for Nest on the OS.

Reviewer 2: ```suggestion
Nest uses a cache to store an offline copy of the list of all available packages for Nest on the OS.
```

Nest can only install packages present in your cache.
That's why you need to **keep it up-to-date** before performing an operation with Nest.

There are 3 different folders for the cache:
* `/var/nest/available` contains the list of all available packages for Nest.
The content of this folder gets updated everytime you update your local cache with Nest.

> * `/var/nest/downloaded` has all the packages that Nest has downloaded, whether they are installed or not.

Reviewer 2: Btw, we will soon add a command to clean that cache. It would be neat, when it's done, to add an example of that command here.

Author: alright, I'll keep that in mind!

If you installed a package before, and then uninstalled it, Nest will keep its archive there, in case you change your mind and want to install that same version of the same package later.
This allows Nest to speed up the process of installation, as it doesn't have to download the same archive again.

> * `/var/nest/installed` contains all the packages that are currently installed on the OS.

Reviewer 2: That is not technically true, it contains information about all the packages currently installed

## 3.3 Version requirement
Sometimes you need a specific version of a package, not just the latest one.
Version requirements solve this problem by allowing you to specify a minimum or maximum version.

> The operators you can use for this are `<`, `<=`, `==`, `>=`, `>` and `*`. The last one simply means 'any' version.

Reviewer 2: There's also `~` and `^`, see this [cheatsheet](https://devhints.io/semver) for more informations =)

The system used for the version of packages is [semver](https://semver.org/).

For example:

> * `stable::kernel/linux#>= 4.1.8` corresponds to a version superior or equal to 4.1.8 of the package `linux` in the category `kernel` of the repository `stable`.

Reviewer 2: ```suggestion
* `stable::kernel/linux#>= 4.1.8` matches any version superior or equal to 4.1.8 of the package `linux` in the category `kernel` of the repository `stable`.
```

> * `dash<0.5.1` correponds to a version strictly inferior to 0.5.1 of the package dash.

Reviewer 2: ```suggestion
* `dash<0.5.1` matches any version strictly inferior to 0.5.1 of the package dash.
```

# 3 Advanced concepts
In this chapter, we will detail some advanced concepts related to `nest`, which will help understanding how `nest` works, but aren't strictly needed for a daily use of `nest`.
Readers interested in `finest` should read this chapter, because those concepts are important to understand how `finest` works.

## 3.1 Managing repositories
### 3.1.1 The official repositories
The official repositories are the ones maintained by the Raven-OS team.
For now, only one is available: `stable`, which is the one enabled by default when installing Nest.
It contains all the stable and safe-to-use packages for Nest, and is maintained by the Raven-OS team.

### 3.1.2 The unofficial repositories
There are also unofficial repositories for Nest which are not vetted by Raven-OS, so caution is advised.
It is up to you to decide whether to trust their maintainers.

### 3.1.3 Configuring repositories

> The repositories configuration is specified in the configuration file of Nest (`/etc/nest/config.toml`).

Reviewer: ```suggestion
The repositories' configuration is specified in the configuration file of Nest (`/etc/nest/config.toml`).
```
I'm not quite certain I got the right meaning, but I think this is the intended one.
```suggestion
The configuration of the repositories is found in the configuration file of Nest (`/etc/nest/config.toml`).
```
(other suggestion)

* `/var/nest/downloaded` has all the packages that Nest has downloaded, whether they are installed or not.
If you installed a package before, and then uninstalled it, Nest will keep its archive there, in case you change your mind and want to install that same version of the same package later.
This allows Nest to speed up the process of installation, as it doesn't have to download the same archive again.
* `/var/nest/installed` contains informations about the packages currently installed on the OS.

## 3.3 Version requirement
Sometimes you need a specific version of a package, not just the latest one.
Version requirements solve this problem by allowing you to specify a minimum or maximum version.

The system used for the version of packages is [semver](https://semver.org/).

The operators you can use for this are `<`, `<=`, `==`, `>=`, `>`, `~`, `^` and `*`:

* The first five of them are pretty self-explanatory.
* The `~` means any version that is superior or equal to the one you specified, but it must have the same major and minor versions.
* The `^` stands for any version that is superior or equal to the one you specified, but it must have the same major version.

> Although there are special cases with `^`:

Reviewer: ```suggestion
However there are special cases with `^`:
```
